### PR TITLE
Decouple dev workflow from Jupyter, create hot debugging environment

### PR DESCRIPTION
**Pull request by Nking92**
Prior to this project, the back end and front end code were both tightly coupled with the Jupyter notebook environment. This project will allow us to run a React app using webpack-dev-server, and run the APIs in a Flask server with debug mode. Both of these toolchains enable hot reloading of components as code is changed, which will significantly improve dev workflow.

Notes about the changes:

- To run the React app, run `npm start` in the project's root directory.
- To run Flask in dev mode, run `flask run` from the project's root directory. The APIs will be hosted at localhost:5050. This configuration is set up in the .flaskenv file.
- Webpack config has been modified to have 2 configs, 1 for dev and 1 for production. The dev config sets up the React hot loader and allows us to run webpack-dev-server to host the app at localhost:3000. 
- widget.css has been moved to the widget/ directory and will be compiled into the widget-build-dev.js file. This will allow the CSS hot loader to work. The style loader module doesn't seem to work with Jupyter, so widget.css will be copied to its original location as a build step as well.
- A VSCode debug configuration has been set up. Press F5 to launch the app in Chrome. The webpack server must be running for it to work. The VSCode debugger should attach and allow you to set breakpoints as usual.
- A bit of code has been moved around so it could be reused to set up models for the dev server.

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -11,6 +13,9 @@ module.exports = {
     new webpack.ProvidePlugin({
       $: "jquery",
       jQuery: "jquery"
+    }),
+    new MiniCssExtractPlugin({
```
Comment (ilmarinen) [inline webpack.config.js:17]: Is this to generate the "widget.css" file within the resources folder? Since you moved the CSS to be included within the React app?

**Review by nushib — APPROVED** [review]

File: webpack.config.js [hunk 2]
```diff
@@ -29,15 +34,83 @@ module.exports = {
...
   },
   output: {
-    path: __dirname + '/backwardcompatibilityml/widget/resources',
+    path: path.resolve(__dirname, 'backwardcompatibilityml/widget/resources'),
     publicPath: '/',
     filename: 'widget-build.js'
   }
 };
+
+const devConfig = {
```
Comment (ilmarinen) [inline webpack.config.js:57]: Do we have a different build target here because we want to inject hot reload?

File: backwardcompatibilityml/widget/app.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) Microsoft Corporation
+# Licensed under the MIT License.
+
+from flask import Flask
+from flask_cors import CORS
+from .compatibility_analysis import init_dev_server
+
+app = Flask(__name__)
+CORS(app)
+init_dev_server(app)
```
[the remark this answers is not in the document]
Comment (ilmarinen) [reply backwardcompatibilityml/widget/app.py:10]: You may need to remove the port specification from the .flaskenv file.

File: webpack.config.js (hunk 2 above)
Comment (ilmarinen) [reply webpack.config.js:57]: If that is the case, would it be possible to keep the same build target, but add the hot reload if we build in development mode?

File: backwardcompatibilityml/widget/resources/index.html [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<!--
```
Comment (ilmarinen) [inline backwardcompatibilityml/widget/resources/index.html:1]: Is there a reason that we cannot re-use the existing widget.html file?

**Review by ilmarinen — CHANGES_REQUESTED** [review]
@Nking92 Really nice work. Thank you.

Overall the approach is good. I'd like to recommend the following structural changes (I've left some comments to suggest how to do them).

1. The current approach works by creating a Flask development app that is separate and parallel to the app that is present within the CompatibilityAnalysis instance.

2. It would be great if we could just re-use this app, and have CompatibilityAnalysis run in development mode, where we can run it as a Flask server from the command line. This shouldn't be too hard. I have left some comments that should help with this.

Besides that, in a few places the development server is referenced as "localhost". It would be great if we could have it listen on "0.0.0.0" instead. I work on a local VM that takes on an IP, and I use the app  / widget by pointing a browser from my host computer to the VM to load the app. So localhost causes things to fail for me. I was able to get things to work by changing this, so you should be able to do it too.

File: webpack.config.js (hunk 1 above)
Comment (jlema) [reply webpack.config.js:17]: +1 on the jupyter stylings - I think the .less files used by jupyter are here: https://github.com/jupyter/notebook/tree/master/notebook/static/notebook/less
In particular, this one seems of interest: https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/less/renderedhtml.less

**Review by jlema — CHANGES_REQUESTED** [review]
+1 on the jupyter stylings - I think the .less files used by jupyter are here: https://github.com/jupyter/notebook/tree/master/notebook/static/notebook/less
In particular, this one seems of interest: https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/less/renderedhtml.less

File: package.json [hunk 5]
```diff
@@ -7,6 +7,7 @@
     "plotly.js-dist": "^1.54.7",
     "react": "^16.7.0",
     "react-dom": "^16.7.0",
+    "react-hot-loader": "^4.13.0",
```
[the remark this answers is not in the document]
Comment (Nking92) [reply package.json:10]: That should be moved to dev dependencies

File: backwardcompatibilityml/widget/resources/index.html (hunk 4 above)
Comment (Nking92) [reply backwardcompatibilityml/widget/resources/index.html:1]: widget.html is a Jinja template, which won't work with `webpack dev server` (unless there's some clever plugin that allows us to do that)

File: webpack.config.js (hunk 2 above)
Comment (Nking92) [reply webpack.config.js:57]: Yes, the general intent is to allow the hot loader to work. The problem I encountered was that the dev mode widget needs some extra stuff compiled into it. What we end up getting are 2 different build artifacts, much like how debug symbols are included in the build with other languages.

`widget-build-dev.js` has the following differences

- React hot loader compiled in
- CSS styles compiled into the JS (which enables hot loading of CSS)

It might be possible to parse command line arguments here and conditionally change up the build depending on that, but that seemed trickier than just having 2 different targets. The current setup will also allow you to test your changes in parallel with Jupyter and the React environment without any need to change up your workflow - just run `npx webpack` to build your latest changes into `widget-build.js`. There won't be any risk of `widget-build.js` and `widget-build-dev.js` conflicting with each other.

Let me know what you think - I'm open to trying to consolidate the build targets if you think it's worthwhile long term.

File: backwardcompatibilityml/widget/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nking92) [reply backwardcompatibilityml/widget/app.py:10]: I'll give that a try. It's probably better for all the initialization code to live here, then go through the CompatibilityAnalysis constructor so we use the same code path.

File: backwardcompatibilityml/widget/resources/index.html (hunk 4 above)
Comment (ilmarinen) [reply backwardcompatibilityml/widget/resources/index.html:1]: Ah, I understand.

Ok, so since we need to run both the webpack dev server and the Python fixture generation code anyway. Could we do this instead:

1. Have the Python fixture generation code, output the HTML (generated from the Junja2 template) and place that in the folder.

2. Have the Webpack dev server run and serve things up from that location.

File: webpack.config.js (hunk 2 above)
Comment (ilmarinen) [reply webpack.config.js:57]: Ok this makes a lot of sense now.

Could we separate out the artifacts into different folders?

So have a top level folder called "development-artifacts" that collect all the build artifacts related to the development server. And drop the `widget-build.js` in there.

Comment (Nking92) [reply webpack.config.js:57]: That's a good idea. Initially I wasn't sure how to separate everything, but that should be doable now.

File: backwardcompatibilityml/widget/resources/index.html (hunk 4 above)
Comment (Nking92) [reply backwardcompatibilityml/widget/resources/index.html:1]: I think we can do that. Are you anticipating needing to inject more things into the template later on? (I noticed "data" is always empty)

File: backwardcompatibilityml/widget/app.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    training_set,
+    testing_set,
+    batch_size_train,
+    batch_size_test,
+    lambda_c_stepsize=0.05,
+    OptimizerClass=optim.SGD,
+    optimizer_kwargs={"lr": learning_rate, "momentum": momentum},
+    NewErrorLossClass=bcloss.BCCrossEntropyLoss,
+    StrictImitationLossClass=bcloss.StrictImitationCrossEntropyLoss,
+    port=5050,
```
Comment (ilmarinen) [inline backwardcompatibilityml/widget/app.py:87]: If you set this to port=5040 and leave the .flaskenv as is ... it all works with dev_mode set to False to skip your workaround.

File: backwardcompatibilityml/widget/compatibility_analysis.py [hunk 7]
```diff
@@ -149,51 +214,19 @@ def __init__(self, folder_name, number_of_epochs, h1, h2, training_set, test_set
...
+            CORS(self.flask_service)
+        else:
+            self.flask_service = FlaskHelper(ip="0.0.0.0", port=port)
+            api_service_environment = build_environment_params(self.flask_service.env)
+            api_service_environment["port"] = self.flask_service.port
+            html_string = render_widget_html(api_service_environment)
+            self.html_widget = HTML(html_string)
+            display(self.html_widget)
+
+        init_app_routes(self.flask_service, self.sweep_manager)
```
Comment (ilmarinen) [inline backwardcompatibilityml/widget/compatibility_analysis.py:232]: Note that in non-dev mode, this actually has to initialize the routes like so:

`init_app_routes(FlaskHelper.app, self.sweep_manager)`

I might have missed that before, if so I apologize.

Comment (Nking92) [reply backwardcompatibilityml/widget/compatibility_analysis.py:232]: Good catch, thanks!

**Comment by Nking92** [comment]
The latest changes fix React hot loading and cleans up the code quite a bit.

- Use HtmlWebpackPlugin to generate index.html at runtime. Webpack will inject the bundled JS code for us. `public/index.html` will be used as a template.
- Removed the `widget-build-dev.js` hack. It's not necessary to output dev build artifacts - webpack dev server actually hosts them in memory. This also fixed React hot loading.
- Combined the 2 webpack configs into 1 by inspecting the `NODE_ENV` environment variable. `NODE_ENV` will be set to `production` when building with `npm run build`.
- `npm run build` is the new way to build instead of `npx webpack`. We need to do it that way to set `NODE_ENV` to production.
- Jupyter notebook styles copied into a style called `.dev-container`. I just inspected the styles with Chrome and extracted the relevant stuff. We won't need to import Jupyter's `.less` files this way.

File: backwardcompatibilityml/widget/resources/index.html (hunk 4 above)
Comment (ilmarinen) [reply backwardcompatibilityml/widget/resources/index.html]: Yeah the injection needs to be cleaned up. It is always possible. Part of the reason for suggesting this is so that we have as minimal differences as possible between how we load the widget in Jupyter vs the dev harness.

File: backwardcompatibilityml/widget/app.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+import torch
+import torch.optim as optim
+import torch.nn.functional as F
+import random
+from backwardcompatibilityml import loss as bcloss
+from backwardcompatibilityml.helpers import training
+from backwardcompatibilityml.helpers.models import LogisticRegression, MLPClassifier
+from rai_core_flask.flask_helper import FlaskHelper
+from .compatibility_analysis import CompatibilityAnalysis
```
Comment (ilmarinen) [inline backwardcompatibilityml/widget/app.py:12]: Can you change this to an absolute import rather than relative.

Comment (ilmarinen) [reply backwardcompatibilityml/widget/app.py:12]: And can we move this file to a different folder, so that it is outside of the widget? Basically so that it doesn't get packaged with the widget.

I'd suggest something like:
```
root_folder
|
|--development
     |
     |--app.py
     |--development-artifacts
     |--public
```

File: backwardcompatibilityml/widget/compatibility_analysis.py [hunk 9]
```diff
@@ -149,51 +213,11 @@ def __init__(self, folder_name, number_of_epochs, h1, h2, training_set, test_set
...
-        @FlaskHelper.app.route("/api/v1/sweep_summary", methods=["GET"])
-        def get_data():
-            return Response(
-                json.dumps(self.sweep_manager.get_sweep_summary()),
-                mimetype="application/json")
-
-        @FlaskHelper.app.route("/api/v1/evaluation_data/<int:evaluation_id>")
-        def get_evaluation(evaluation_id):
-            return self.sweep_manager.get_evaluation(evaluation_id)
+        init_app_routes(FlaskHelper.app, self.sweep_manager)
```
Comment (ilmarinen) [inline backwardcompatibilityml/widget/compatibility_analysis.py:223]: Can you move this up so that this happens right after the Flask service is instantiated.

File: package.json [hunk 10]
```diff
@@ -28,11 +29,21 @@
...
+    "style-loader": "^2.0.0",
     "ts-loader": "^8.0.1",
     "typescript": "^3.9.7",
     "webpack": "^4.43.0",
     "webpack-cli": "^3.2.1",
-    "webpack-dev-server": "^3.1.14"
+    "webpack-dev-server": "^3.11.0"
+  },
+  "scripts": {
+    "start": "webpack-dev-server --hot --mode development",
```
Comment (ilmarinen) [inline package.json:46]: Can you add `--host 0.0.0.0` so that it is not listening on localhost by default.
Recall our conversation regarding my dev setup where its running in a VM with a specific IP so that I'm not running against localhost.

File: public/index.html [hunk 11]
```diff
@@ -0,0 +1,36 @@
+<!--
```
Comment (ilmarinen) [inline public/index.html:1]: Can't we re-use the existing Jinja2 template by having the Python app bootstrapping process just render the template as this file?

File: widget/widget.css [hunk 12]
```diff
@@ -3,6 +3,15 @@
...
 
+ .dev-container {
+    box-sizing: border-box;
+    color: #010;
+    font-size: 14px;
+    line-height: 20px;
+    font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
+    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
+ }
+
```
Comment (ilmarinen) [inline widget/widget.css:14]: So I did a visual comparison between what the widget looks like in the dev container vs Jupyter, and I still see differences.

Attaching two images. First one is within Jupyter and second one is within the dev window:

Screen Shot 2020-10-14 at 11 09 05 PM

Screen Shot 2020-10-14 at 11 21 36 PM

Notice how some of the spacing is missing?

**Review by ilmarinen — COMMENTED** [review]
@Nking92 Looks good. Thank you for making the changes requested so that the development version of the code is almost identical to how we would run the widget within Jupyter. Great job there.

The one lingering point where we differ from the widget is in the fact that we are using two template files. But I suspect that this is needed in order for hot reload to work? If so then I am willing to accept this as a necessary evil. We will just need to remember to keep these files in sync and document that fact somewhere. But if it is not needed then lets try and use just one file.

I am still noticing some differences in rendering between the dev window and within Jupyter. I have attached two images in a comment. Let me know if you are seeing the same thing.

Besides that, I have left a few structural comments regarding where files should live etc. this is minor stuff that needs to be done to keep the structure tidy.

Great job. Thank you!

**Comment by ilmarinen** [comment]
Oh one thing that I forgot to mention:

Just an FYI that you can still build with `npx webpack`, you just need to do `NODE_ENV=production npx webpack`.

But in either case, would you mind documenting the steps for:

1. Building the widget and give both alternatives, i.e. `npm run build` and the adjustment I mentioned above.
2. Detailed instructions for running and using the dev harness.

You can add these to the README.


Thank you!

File: widget/widget.css (hunk 12 above)
Comment (Nking92) [reply widget/widget.css:14]: Thanks for the screenshot. I wasn't sure exactly what else was missing. I'll see if I can fix that.

**Review by jlema — APPROVED** [review]

File: .flaskenv [hunk 13]
```diff
@@ -0,0 +1,2 @@
+FLASK_APP=development/app.py
+FLASK_ENV=development
```
Comment (ilmarinen) [inline .flaskenv:2]: You can add the `FLASK_RUN_PORT=5000` back here, since we're using that in the API and need it to be the same. Othwerwise there may be a time when it chooses some other random port.

File: README.md [hunk 14]
```diff
@@ -41,6 +41,16 @@ To run tests, make sure that you are in the project root folder and do:
 1. `pip install -r dev-requirements.txt`
 2. `pytest tests/`
 
+# Development environment
+
+To run the widget and APIs in a development environment:
+1. Open a new terminal, then from within the root folder do `npm start`. This will host the widget with `webpack-dev-server`.
+2. Open a new terminal, then from within the root folder do `flask run`. This will start the Flask server for the APIs used by the widget.
```
Comment (ilmarinen) [inline README.md:48]: Can you also specify options for how one should run this for when you want to develop against an ip other than localhost?

File: README.md [hunk 15]
```diff
@@ -41,6 +41,16 @@ To run tests, make sure that you are in the project root folder and do:
...
+
+The widget can be loaded in the web browser at `localhost:3000`. It will be loaded independently from a Jupyter notebook. The APIs will be hosted at `localhost:5000`.
```
Comment (ilmarinen) [inline README.md:50]: Some words about the non-localhost case as well should go here.

File: backwardcompatibilityml/helpers/training.py [hunk 16]
```diff
@@ -827,7 +827,7 @@ def compatibility_sweep(sweeps_folder_path, number_of_epochs, h1, h2,
 
     sweep_summary = {
         "data": sweep_summary_data,
-        "h1_performance": model_accuracy(h1, test_set, device=device)
```
Comment (ilmarinen) [inline backwardcompatibilityml/helpers/training.py:830]: Why is this change here?

**Review by ilmarinen — APPROVED** [review]
@Nking92 +1
